stl/inc/format

```diff
@@ -3934,6 +3938,251 @@ _NODISCARD size_t formatted_size(const locale& _Loc, const wformat_string<_Types
+                }
+                _STD get<_Idx>(_Underlying).format(_Arg, _Tmp_ctx);
+            };
+            (_Single_writer.template operator()<_Indice>(_Args), ...);
+        }(index_sequence_for<_ArgTypes...>{});
+        _STD _Copy_unchecked(_Closing_bracket._Unchecked_begin(), _Closing_bracket._Unchecked_end(), _Tmp_ctx.out());
+
+        return _STD _Write_aligned(_Fmt_ctx.out(), static_cast<int>(_Tmp_buf.size()), _Format_specs, _Fmt_align::_Left,
+            [&](typename _FormatContext::iterator _Out) {
+                return _STD _Fmt_write(_STD move(_Out), basic_string_view<_CharT>{_Tmp_buf});
+            });
+    }
+
+public:
+    constexpr void set_separator(const basic_string_view<_CharT> _Sep) noexcept {
+        _Separator = _Sep;
+    }
+
+    constexpr void set_brackets(
+        const basic_string_view<_CharT> _Opening, const basic_string_view<_CharT> _Closing) noexcept {
+        _Opening_bracket = _Opening;
+        _Closing_bracket = _Closing;
+    }
+
+    template <class _ParseContext>
+    constexpr _ParseContext::iterator parse(_ParseContext& _Ctx) {
+        _Fmt_tuple_type _Fmt_type = _Fmt_tuple_type::_None;
+        _Tuple_format_specs_setter<_CharT, sizeof...(_Types) == 2> _Callback{_Specs, _Fmt_type, _Ctx};
+        const auto _It = _STD _Parse_tuple_format_specs(_Ctx._Unchecked_begin(), _Ctx._Unchecked_end(), _Callback);
+        if (_It != _Ctx._Unchecked_end() && *_It != '}') {
+            _STD _Throw_format_error("Missing '}' in format string.");
+        }
+
+        if (_Fmt_type == _Fmt_tuple_type::_No_brackets) {
+            set_brackets({}, {});
+        } else if constexpr (sizeof...(_Types) == 2) {
+            if (_Fmt_type == _Fmt_tuple_type::_Key_value) {
+                set_separator(_STATICALLY_WIDEN(_CharT, ": "));
+                set_brackets({}, {});
+            }
+        }
+
+        _Ctx.advance_to(_Ctx.begin() + (_It - _Ctx._Unchecked_begin()));
+        _STD apply([&_Ctx](auto&&... _Elems) { (_Tuple_debug_format_setter(_Elems, _Ctx), ...); }, _Underlying);
+
+        return _Ctx.begin();
+    }
+};
+
+// formatter definition for all pairs and tuples, the deleted default constructor
+// makes it "disabled" as per N4971 [format.formatter.spec]/5
+
+template <class _TupleOrPair, class _CharT>
+struct _Tuple_formatter_base {
+    _Tuple_formatter_base()                                       = delete;
+    _Tuple_formatter_base(const _Tuple_formatter_base&)           = delete;
+    _Tuple_formatter_base operator=(const _Tuple_formatter_base&) = delete;
+};
+
+template <class _CharT, formattable<_CharT> _Ty1, formattable<_CharT> _Ty2>
+struct _Tuple_formatter_base<pair<_Ty1, _Ty2>, _CharT> : _Tuple_formatter_common_base<_CharT, _Ty1, _Ty2> {
+private:
+    using _Formatted_type =
+        _Maybe_const<_Tuple_formatter_common_base<_CharT, _Ty1, _Ty2>::_Is_const_formattable, pair<_Ty1, _Ty2>>;

```

Comment:
Style: If these are going to wrap, we may as well introduce `_Base` typedefs to make this more readable.